### ISSUE TITLE
request for clarification: why does kube-aws update take down my controller

### ISSUE THREAD

**Issue by diffalot** [coreos/kube-aws#50]
I am very new to kubernetes and coreos, but i'm confused about why the following commands take down my cluster for a few minutes (on v0.9.1):

```
# change workerCount in cluster.yaml from 1 to 2
kube-aws render stack
kube-aws validate
kube-aws update
```

I was expecting another worker node to be launched.

**Comment by mumoshu** [comment]
@diffalot Hi, thanks for trying out kube-aws!

That's because we rely on the `AutoScalingRollingUpdate` setting of CloudFormation as described in [the AWS dochttp://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html] and implemented [like](https://github.com/coreos/kube-aws/blob/master/config/templates/stack-template.json#L109-L115) [this](https://github.com/coreos/kube-aws/blob/master/config/templates/stack-template.json#L53-L62).

If you'd need the way that works right now, as we have `MaxBatchSize` of 1, I suggest you to have 2 or more workers/controllers to avoid downtime so that ASG will rolling update instances one by one.

**Comment by mumoshu** [comment]
Oh, is what you've changed is just workerCount, right?
If thats the case, at least it isn't an expected behavior.

**Comment by mumoshu** [comment]
@diffalot OK now I guess I've spotted the issue.
Here's what we can see when you've changed workerCount from 4 to 5, controllerCount from 2 to 3.

![image](https://cloud.githubusercontent.com/assets/22009/20162465/9a54d2bc-a738-11e6-9c3e-bee512b4cd73.png)

According to the events recorded in cloudformation, kube-aws of today does always involve `rolling upgrade` of instances, even though it is just a change to workerCount/controllerCount (!).
Also, kube-aws teaches ASG to wait for predefined period of time until it assumes new worker/controller to be ready to serve your workloads. This, in combination with the fact mentioned above, will result in some downtime when you have just 1 worker/controller and the new one replacing it unfortunately don't come up before the predefined period elapsed.

In short term, suggestions:

* Make the pause time longer by editing your stack-template.json, and/or
* Choose larger instance size for your worker/controller, so that it comes up fast enough (before the predefined period=pause time elapses)
* Have 2 or more worker/controller, so single-non-ready worker/controller won't incur downtime

In longer term:

* There is already an proposal which will eventually fix kube-aws to wait until worker/controller to be actually ready, surprisingly (see https://github.com/coreos/kube-aws/issues/49)

**Comment by diffalot** [comment]
ah ha, I think i'm missing the `controllerCount` variable in the cluster.yml and I'm guessing it has something to do with starting out with `kube-aws init` on v0.8.x and then switching to v0.9.x so i could try out the `kube-aws update` command.

thank you very much, @mumoshu, you've given me a very good lead to investigate further.

I'll keep playing around with it and report back here.

Thank you all for putting kube-aws together, it really is a wonderful piece of software.  And thanks for helping out a newb trying to leverage the system :)

**Comment by mumoshu** [comment]
@diffalot Ah, I didn't guess that! Thanks for clarifying your use case! It really helps.

Yes, your case would also result in some downtime, because I believe we didn't explicitly prioritize resources creation/update order in that case(v0.8.x kube-aws init --> v0.9.x kube-aws update).

Also thanks for the kind words. I wish you'll be happy with kube-aws!
Let me know if anything 👍

**Comment by diffalot** [comment]
I just want to leave a note hear that may provide clarity for other newcomers to kubernetes who are spinning up `kube-aws` as a trial. (may be relevant for #138)

1. `kube-aws` is awesome!  it's such a forgiving environment that it's a little bit scary.  I've messed up more things than I can count, and the the solution is inevitably, run `kube-aws render stack && kube-aws validate && kube-aws update`.  It's like a worm drive for running apps.
1. definitely run an `m3.medium` controller and at least 3 workers
1. Always run 2 copies of every app, if you can.  This will insure that when you update your cluster, your apps will stay deployed 100% while the new controller is rolling them over to the new workers.

Thanks for the great software, and please correct me if any of my experiences are against best practices.
